src/main/java/com/segment/analytics/android/integrations/firebase/FirebaseIntegration.java

```diff
@@ -3,88 +3,297 @@
+          trait = trimKey(trait);
+        }
+
+        String formattedValue;
+
+        if (entry.getValue() instanceof Date) {
+
+          Date value = (Date) entry.getValue();
+          formattedValue = formatDate(value);
+
+        } else if (entry.getValue() instanceof String && ISO_DATE.matcher(String.valueOf(entry.getValue())).matches()) {
+
+          formattedValue = String.valueOf(entry.getValue()).substring(0, 10);
+
+        } else {
 
+          formattedValue = String.valueOf(entry.getValue());
+        }
+
+        mFirebaseAnalytics.setUserProperty(trait, formattedValue);
+        logger.verbose("mFirebaseAnalytics.setUserProperty(%s, %s);", trait, formattedValue);
+      }
+    }
   }
 
-  @Override public void screen(ScreenPayload screen) {
+  @Override
+  public void screen(ScreenPayload screen) {
     super.screen(screen);
-//    mFirebaseAnalytics.setCurrentScreen(this, screen.name(), null /* class override */);
+
+    logger.verbose(
+            "The Firebase SDK gathers screen calls automatically. Segment does map manual screen" +
+                    "methods to Firebase.");
+
   }
 
-  @Override public void track(TrackPayload track) {
+  @Override
+  public void track(TrackPayload track) {
     super.track(track);
 
+    String event = track.event();
+    String eventName = mapEvent(event);
+
+    Properties properties = track.properties();
+    Bundle formattedProperties = formatProperties(properties);
+
+    mFirebaseAnalytics.logEvent(eventName, formattedProperties);
+    logger.verbose("mFirebaseAnalytics.logEvent(%s, %s);", eventName, formattedProperties);
+  }
+
+  private String mapEvent(String event) {
+
+    String eventName = event;
+
+    final Map<String, String> eventMapper = new HashMap<>();
+    eventMapper.put("Product Added", Event.ADD_TO_CART);
+    eventMapper.put("Checkout Started", Event.BEGIN_CHECKOUT);
+    eventMapper.put("Order Completed", Event.ECOMMERCE_PURCHASE);
+    eventMapper.put("Order Refunded", Event.PURCHASE_REFUND);
+    eventMapper.put("Product Viewed", Event.VIEW_ITEM);
+    eventMapper.put("Product List Viewed", Event.VIEW_ITEM_LIST);
+    eventMapper.put("Payment Info Entered", Event.ADD_PAYMENT_INFO);
+    eventMapper.put("Promotion Viewed", Event.PRESENT_OFFER);
+    eventMapper.put("Product Added to Wishlist", Event.ADD_TO_WISHLIST);
+    eventMapper.put("Product Shared", Event.SHARE);
+    eventMapper.put("Product Clicked", Event.SELECT_CONTENT);
+    eventMapper.put("Product Searched", Event.SEARCH);
+    eventMapper.put("Promotion Viewed", Event.PRESENT_OFFER);
+
+    if (eventMapper.containsKey(eventName)) {
+      eventName = eventMapper.get(eventName);
+    }
+
+    if (eventName.contains(" ")) {
+      eventName = eventName.trim().toLowerCase().replaceAll(" ", "_");
+    }
+
+    if (eventName.length() > 40) {
+      eventName = trimKey(eventName);
+    }
+
+    return eventName;
+  }
+
+  private Bundle formatProperties(Properties properties) {
+
```

Comment:
Unnecessary blank line as per https://google.github.io/styleguide/javaguide.html#s4.6.1-vertical-whitespace. Please take a second pass through the file to see if you missed any others. I'm still seeing a few others.